docs/migrating.rst

```diff
@@ -0,0 +1,235 @@
+######################################
+Migrating from Python 2 version of NDB
+######################################
+
+While every attempt has been made to keep compatibilty with the previous
+version of `ndb`, there are fundamental differences at the platform level,
+which have made necessary in some cases to depart from the original
+implementation, and sometimes even to remove exisitng functionality
+altogether.
+
+Because one of the main objectives of this rewrite was to be able to use `ndb`
+independently from Google App Engine, the legacy APIs from GAE cannot be
+depended upon. Also, any environment and runtime variables and resources will
+not be available when running outside of GAE. This means that many `ndb` APIs
+that depended on GAE have been changed, and many APIs that accessed GAE
+resources directly have been dropped.
+
+Aside from this, there are many differences between the Datastore APIs
+provided by GAE and those provided by the newer Google Cloud Platform. These
+diffeences have required some code and API changes as well.
+
+Finally, in many cases, new features of Python 3 have eliminated the need for
+some code, particularly from the old `utils` module.
+
+If you are migrating code, these changes can generate some confusion. This
+document will cover the most common migration issues.
+
+Setting up a connection
+=======================
+
+The most important difference from the previous `ndb` version, is that the new
+`ndb` requires the use of a client to set up a runtime context for a project.
+This is necessary because `ndb` can now be used in any Python environment, so
+we can no longer assume it's running in the context of a GAE request.
+
+The `ndb` client uses ``google.auth`` for authentication, which is how APIs in
+Google Cloud Platform work. The client can take a `credentials` parameter or
+get the credentials using the `GOOGLE_APPLCATION_CREDENTIALS` environment
+variable, which is the recommended option.
+
+After instantiating a client, it's necessary to establish a runtime context,
```

Comment:
Mention that the context is not thread safe and should be used one per request if being used in a web application, for example.